### PR TITLE
Migrate all stages to StageExample + pipeline_tester + auto stage docs

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary

Second in the sequence (PR #79 bumped `MEDS-transforms` to 0.6.7). Migrates **all 8 MEDS_extract stages** onto MEDS-transforms 0.6.x's `StageExample`-based testing machinery and adds auto-generated stage docs — the same pattern adopted by [meds-torch-data#86](https://github.com/mmcdermott/meds-torch-data/pull/86) and [meds-torch-data#90](https://github.com/mmcdermott/meds-torch-data/pull/90).

Replaces the earlier `finalize_MEDS_data`-only pilot on this same branch and does the full migration in one PR, since the remaining stages require a custom `StageExample` subclass (upstream only handles `MEDSDataset`-formatted I/O, which most MEDS_extract stages don't produce).

## Infrastructure

**`src/MEDS_extract/_stage_example.py`** — new `MEDSExtractStageExample(StageExample)` subclass:

- `is_example_dir` / `from_dir` treat `in.yaml`, `out_data.yaml`, `out_metadata.yaml` as `yaml_to_disk` Path specs (not parsed MEDSDatasets). Optional `cfg.yaml` → `stage_cfg`, `pipeline_cfg.yaml` → `pipeline_cfg`, `_test_cfg.yaml` → constructor kwargs.
- `check_outputs` materializes the expected tree via `yaml_disk` into a temp dir and compares each file to the stage's actual output by suffix: `.parquet` via `polars.testing.assert_frame_equal` (with `check_row_order=False` as a package-level default since rwlock_wrap shuffles + multi-file concats make outputs inherently non-deterministic), `.json` / `.yaml` / `.yml` via parsed-dict equality (order-independent for hashable-list values, so e.g. `.shards.json` doesn't depend on split-iteration order).
- In pipeline mode, falls back from `stage_output_dir` to `output_dir.parent` (cohort root) when a file isn't at the stage's own dir — covering globally-shared pipeline outputs like `shards_map_fp: ${output_dir}/metadata/.shards.json`.
- Skips `.logs/` / `.hydra/` dirs and known byproducts (`event_conversion_config.yaml`, `dataset.json`).
- `render_content` override handles `want_metadata: Path` for auto-docgen (the upstream default calls `df_to_markdown`, which blows up on a path).

Every stage now registers with `@Stage.register(..., example_class=MEDSExtractStageExample)`. `Stage.register` auto-infers `examples_dir` from the sibling directory so no extra wiring is needed.

Upstream discussion of what could be hoisted: mmcdermott/MEDS_transforms#387.

## Per-stage examples

| Stage | Scenarios | Expected output |
|---|---|---|
| `shard_events` | `default` | `data/<prefix>/[start-end).parquet` row-chunks |
| `split_and_shard_subjects` | `default`, `join` | `metadata/.shards.json` |
| `convert_to_subject_sharded` | `default`, `join` | `data/<split>/<shard>/<table>.parquet` |
| `convert_to_MEDS_events` | `default` | MEDS-format event parquets with `code_components` + `source_block` |
| `merge_to_MEDS_cohort` | `default` | Merged `data/<split>/<shard>.parquet` |
| `extract_code_metadata` | `default` | `metadata/codes.parquet` with `code_template` |
| `finalize_MEDS_metadata` | `default` | `metadata/codes.parquet`, `metadata/subject_splits.parquet` |
| `finalize_MEDS_data` | `default` | Schema-aligned MEDS data parquets |

The `join` scenarios on `convert_to_subject_sharded` and `split_and_shard_subjects` exercise `_table.join` — replacing the coverage previously in `tests/test_join_tables.py`.

## Tests

- `tests/test_stage_examples.py` — one-line pytest-plugin entry point; auto-parametrizes across all `(stage, scenario)` pairs. **10 stage-example tests.**
- `tests/test_pipeline.py` — `pipeline_tester`-based end-to-end run linking all 8 stages' `default` scenarios in sequence. **1 pipeline test.**
- `tests/test_config.py` — absorbed the 4 library-level `needed_source_columns` regression tests that used to live in `tests/test_shard_events.py`.
- **Deleted**: `test_shard_events.py`, `test_split_and_shard_subjects.py`, `test_convert_to_subject_sharded.py`, `test_convert_to_MEDS_events.py`, `test_merge_to_MEDS_cohort.py`, `test_extract_code_metadata.py`, `test_finalize_MEDS_metadata.py`, `test_join_tables.py`. `tests/utils.py` pruned from 487 → 46 lines (only `assert_df_equal` kept; still referenced by `test_inprocess_pipeline.py`). `tests/__init__.py` emptied.

## Auto stage docs

- `docs/gen_stage_docs.py` — near-verbatim copy of the upstream / MTD pattern, pointed at `"MEDS_extract"`.
- `mkdocs.yml` — registers the new script under `gen-files` scripts and adds a top-level `Stages` nav entry.
- `.gitignore` — adds `docs/api/` and `docs/stages/` so local builds don't leak artifacts.

`uv run mkdocs build` produces 8 stage pages at `site/stages/<stage_name>/index.html` — each rendering the stage docstring, metadata, default config, schema updates, CLI usage, and per-scenario `in.yaml` / `out_data.yaml` / `out_metadata.yaml` code blocks.

## Dependencies

- `yaml_to_disk>=0.0.4` promoted from dev to a runtime dependency (needed at both test time and by any downstream consumer that wants to run the packaged examples).
- `[tool.setuptools.package-data]` ships `**/examples/**/*.{yaml,md}` in the wheel.

## Net impact

~2.3k LOC of imperative Python test scaffolding traded for ~1.2k LOC of declarative YAML examples + 281 LOC of shared helper — roughly **−50%** shrink plus a qualitative shift from "read Python" to "read YAML". Full assessment in PR thread.

## Test plan

- [x] `uv run pytest` — 66 passed (10 stage examples + 1 pipeline + 27 in-process edge cases + 6 config regressions + 22 library doctests)
- [x] `uv run mkdocs build` — 8 stage pages produced
- [x] `uv run ruff check` clean
- [x] `uv build --wheel` — example YAMLs ship in the wheel

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/MEDS_extract/_stage_example.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        case ".parquet":
+            kwargs = {"check_column_order": False, "check_dtypes": False, **df_check_kwargs}
+            # `glob=False` because shard_events names files `[start-end).parquet` — the
+            # brackets are glob metacharacters in polars' default reader.
+            got = pl.read_parquet(actual_fp, glob=False)
+            want = pl.read_parquet(expected_fp, glob=False)
+            try:
+                assert_frame_equal(
+                    got.sort(want.columns) if set(got.columns) == set(want.columns) else got,
+                    want.sort(want.columns),
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_extract/_stage_example.py]: Good catch — fixed in d92cef7. Removed the local `sort(want.columns)` hack and replaced it with polars' native `check_row_order=False` kwarg (polars sorts both frames internally when that's set, so the behavior for the non-deterministic cases is preserved but callers retain an off-switch).

`check_row_order: False` is now a package-level default in `_compare` because MEDS_extract stages pass input through `rwlock_wrap` (which shuffles shards) and concatenate multiple source sub-shards non-deterministically — so strict-order comparisons would be flaky across most scenarios. Callers that *do* need ordering-sensitivity can opt in via `_test_cfg.yaml`:

```yaml
# example/_test_cfg.yaml
df_check_kwargs:
  check_row_order: true
```

The default-then-override pattern also survives `pipeline_tester`'s clobber of `df_check_kwargs` (it sets `rel_tol`/`abs_tol`/`check_dtypes` but not `check_row_order`, so our package default wins), which was the reason the sort hack existed in the first place.

File: tests/test_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+"""
+
+from MEDS_transforms.pytest_plugin import pipeline_tester
+
+PIPELINE_YAML = """
+input_dir: "{input_dir}"
+output_dir: "{output_dir}"
+event_conversion_config_fp: "{input_dir}/event_cfg.yaml"
+shards_map_fp: "{output_dir}/split_and_shard_subjects/.shards.json"
+etl_metadata:
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_pipeline.py:15]: Valid — fixed in d92cef7. The non-canonical path was a workaround: `pipeline_tester` calls `stage.check_outputs(stage_output_dir={cohort}/<stage_name>, is_resolved_dir=True)` for non-last metadata stages, so with canonical `shards_map_fp: ${output_dir}/metadata/.shards.json` the file lived OUTSIDE the stage's own output dir and `check_outputs` couldn't find it.

Restored the canonical path and fixed `MEDSExtractStageExample.check_outputs` properly: when `is_resolved_dir=True` and a file isn't found at the stage's own dir, it now falls back to `output_dir.parent` (the cohort root) using the full un-stripped relative path — exactly matching the layout that globally-shared pipeline outputs like `shards_map_fp` actually use.

The test now exercises the real pipeline layout, including the fact that `split_and_shard_subjects`' `.shards.json` output writes to the pipeline-global path rather than the stage's own dir.

File: tests/test_stage_examples.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+"""Run every registered MEDS_extract stage example end-to-end.
+
+The `stage_example` fixture is auto-parametrized by `MEDS_transforms.pytest_plugin` across all
+`(stage, scenario)` pairs in the `MEDS_extract` package, so this file only needs to stand up one
+test function. A stage migrates onto this machinery by placing `in.yaml` / `out_data.yaml` under
+`<stage>/examples/` (with optional nested `<scenario>/` subdirectories); `Stage.register`
+auto-infers `examples_dir` from the stage's own directory.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_stage_examples.py]: Fixed in af9b991. Docstring now mentions both `out_data.yaml` and `out_metadata.yaml`:

> A stage migrates onto this machinery by placing `in.yaml` plus either `out_data.yaml` (for data stages) or `out_metadata.yaml` (for metadata stages) under `<stage>/examples/<scenario>/`.

File: src/MEDS_extract/_stage_example.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+            raise AssertionError(f"Unsupported output suffix '{expected_fp.suffix}' for {rel}.")
+
+
+def _assert_struct_equal(got: object, want: object, rel: Path) -> None:
+    """Compare two structured blobs (dicts/lists), treating top-level dict values as sets when they are lists
+    of hashables.
+
+    Motivated by ``.shards.json`` — ``split_and_shard_subjects`` output where per-split subject
+    lists aren't deterministically ordered run-to-run. Nested dicts recurse; anything else is a
+    strict ``==`` comparison.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_extract/_stage_example.py]: Fixed in af9b991 by updating the docstring (option 2 from your suggestion). The recursive unordered-list behavior is actually the more useful semantics — a future pipeline output with e.g. `{split: {shard: [subjects]}}` nesting should still be compared order-independently at the inner list level — so I kept the recursion and made the docstring match:

> Compare two structured blobs (dicts/lists) **recursively**, treating dict values as unordered sets whenever they are lists of hashables at **any** nesting level.

File: tests/test_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+"""End-to-end pipeline test via ``MEDS_transforms.pytest_plugin.pipeline_tester``.
+
+Runs all eight MEDS_extract stages in sequence and validates each stage's output against its
+registered ``default`` scenario. The first stage's ``in.yaml`` bootstraps the pipeline's input
+(raw CSVs + event_cfg.yaml); each subsequent stage reads the prior stage's output.
+"""
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_pipeline.py:6]: Good catch. The PR body was still the pilot version — my earlier `gh pr edit --body` attempts silently failed on unrelated "Projects (classic) deprecated" GraphQL errors and left the old description in place. Just rewrote it via the REST API (`PATCH /repos/:owner/:repo/pulls/80`), which worked. The body now correctly describes the full migration (all 8 stages + pipeline test + auto stage docs) with no misleading "follow-up PR" references for items that actually land here.

File: tests/test_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      row_chunksize: 2
+      infer_schema_length: 10000
+  - split_and_shard_subjects:
+      split_fracs:
+        train: 0.5
+        tuning: 0.25
+        held_out: 0.25
+      n_subjects_per_shard: 10
+      seed: 1
+  - convert_to_subject_sharded
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/test_pipeline.py:29]: Verified and fixed in e943d4a. `split_and_shard_subjects.py:263` reads `cfg.seed` (pipeline-level), so the nested `seed: 1` under the stage block was indeed a no-op — the test passed only by coincidence that `MEDS_transforms.configs._main.yaml` defaults `seed: 1` too. Moved `seed: 1` to the top level of `PIPELINE_YAML` where the code actually reads from:

```yaml
etl_metadata:
  dataset_name: TEST
  dataset_version: "0.1"
seed: 1
stages:
  - split_and_shard_subjects:
      split_fracs: {...}
      n_subjects_per_shard: 10
      # no more stage-level 'seed'
```

File: src/MEDS_extract/split_and_shard_subjects/examples/default/cfg.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+split_fracs:
+  train: 0.5
+  tuning: 0.25
+  held_out: 0.25
+n_subjects_per_shard: 10
+seed: 1
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_extract/split_and_shard_subjects/examples/default/cfg.yaml]: Fixed in e943d4a. Moved `seed: 1` from `cfg.yaml` (stage_cfg) into `pipeline_cfg.yaml` where the stage actually reads it from:

```yaml
# pipeline_cfg.yaml
event_conversion_config_fp: ${input_dir}/event_cfg.yaml
shards_map_fp: ${output_dir}/metadata/.shards.json
seed: 1
```

Same fix applied to the `join` scenario ([see related thread](https://github.com/mmcdermott/MEDS_extract/pull/80#discussion_r3125664902)).

File: src/MEDS_extract/split_and_shard_subjects/examples/join/cfg.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+split_fracs:
+  train: 0.5
+  tuning: 0.5
+  held_out: null
+n_subjects_per_shard: 10
+seed: 1
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/MEDS_extract/split_and_shard_subjects/examples/join/cfg.yaml]: Same fix as the default scenario — moved `seed: 1` from `cfg.yaml` to `pipeline_cfg.yaml` in e943d4a.

File: pyproject.toml [hunk 9]
```diff
@@ -35,14 +36,16 @@ dev = [
...
-    "yaml_to_disk>=0.0.4",
 ]
 docs = [
   "mkdocs==1.6.1", "mkdocs-material==9.6.7", "mkdocstrings[python,shell]==0.28.2", "mkdocs-gen-files==0.5.0",
   "mkdocs-literate-nav==0.6.1", "mkdocs-section-index==0.3.9", "mkdocs-git-authors-plugin==0.9.2",
   "mkdocs-git-revision-date-localized-plugin==1.3.0"
 ]
 
+[tool.setuptools.package-data]
+MEDS_extract = ["**/examples/**/*.yaml", "**/examples/**/*.md"]
```
Comment (mmcdermott) [inline pyproject.toml]: Why is this necessary? These aren't used in prod, right?

File: tests/utils.py [hunk 10]
```diff
@@ -1,200 +1,14 @@
...
-            conf_dir.cleanup()
-        raise AssertionError(
-            f"{test_name} failed as command errored when not expected!\n" + "\n".join(err_cmd_lines)
-        )
-    if do_use_config_yaml:
-        conf_dir.cleanup()
-    return stderr, stdout
 
 
 def assert_df_equal(
```
Comment (mmcdermott) [inline tests/utils.py]: Do we actually still need this?

File: src/MEDS_extract/_stage_example.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+import polars as pl
+from MEDS_transforms.stages.examples import StageExample
+from omegaconf import OmegaConf
+from polars.testing import assert_frame_equal
+from yaml_to_disk import yaml_disk
+
+
+@dataclass
+class MEDSExtractStageExample(StageExample):
+    """StageExample variant that reads inputs/outputs as ``yaml_to_disk`` specs on disk.
```
Comment (mmcdermott) [inline src/MEDS_extract/_stage_example.py:36]: Some doctests here would help resolve our missing code coverage lines significantly, I bet.

File: pyproject.toml (hunk 9 above)
Comment (mmcdermott) [reply pyproject.toml]: You're right — they're not. Reverted in 3c43a6b:

- Dropped `[tool.setuptools.package-data]` rule for `examples/**/*.{yaml,md}`.
- Reverted `yaml_to_disk` from runtime to dev-only (it was only used by the stage-example runtime/test layer).

The examples are needed by (a) `pytest test_stage_examples.py` (dev time, reads from source tree) and (b) `mkdocs build` (source tree). Downstream consumers importing a MEDS_extract stage never re-run or re-render the examples, so no reason to carry them in the wheel.

File: tests/utils.py (hunk 10 above)
Comment (mmcdermott) [reply tests/utils.py]: Good catch — no. `assert_df_equal` had exactly one caller left (`test_assert_df_equal_detects_extra_columns`), which was a self-regression test verifying the helper's own behavior. Both deleted in 3c43a6b; `tests/utils.py` is gone entirely now.

File: src/MEDS_extract/_stage_example.py (hunk 11 above)
Comment (mmcdermott) [reply src/MEDS_extract/_stage_example.py:36]: Done in 3c43a6b. Added doctests on `MEDSExtractStageExample`, `render_content`, `_compare`, `_assert_struct_equal`, and `_is_hashable_list` covering:

- All `__post_init__` validation branches (both-None / both-not-None errors, scenario_name canonicalization, pipeline_cfg auto-sets-config-yaml, default df_check_kwargs).
- `render_content` with `want_metadata: Path` (both the Path + DataFrame branches).
- All four `_compare` suffix dispatches (parquet, json, yaml, unsupported) with match + mismatch cases each.
- `_assert_struct_equal`: order-independent hashable list values, missing/extra key errors, contents-differ errors, nested-dict recursion, non-dict scalar mismatch.
- `_is_hashable_list`: lists of hashables, empty lists, lists-of-lists (unhashable), non-lists, dicts.

Coverage on `_stage_example.py` jumps from **76% → 98%**; overall project from 92% → 95%. Three lines remain uncovered (`check_outputs`' missing-file fallback + error raise paths) — those require a full on-disk output dir to exercise, and the `pytest test_stage_examples.py` happy-path doesn't reach them; leaving them since the coverage target was "significantly help" and the remaining 3 are error-raising internals, not behavior paths.
